feat: make nats memory limits configurable

domdom82 opened the pull request:
Depending on the deployment size of Cloud Foundry, 500 MB may not be enough memory for NATS.
We ran into issues where BOSH was restarting NATS VMs just because they reported this alert, even when there where no issues.

We would like to make the limits configurable, yet keep the current values as defaults so that this should be a no-op for everyone else.

- Remove the hard-coded memory limit
- Replace with with spec properties
- Keep current values as default

domdom82 committed:
feat: make nats memory limits configurable

geofffranks reviewed (changes requested):
Can you add template tests for the monit file in `spec/`? 

What happens if the string format isn't followed, or is invalid - does this break monit just the `nats` process, everything managed by monit on the VM, silently fail, or complain loudly in a way operators can track down easily?

domdom82 commented:
tried setting the limit to `500 potatoes` as a test.
BOSH agent does not like it at all:
```
Task 231250 | 11:19:17 | L starting jobs: nats/e936cbd2-8d46-4099-bf4e-359be48d240c (0) (canary) (00:01:15)
                       L Error: Timed out sending 'start' to instance: 'nats/e936cbd2-8d46-4099-bf4e-359be48d240c', agent-id: 'bc8584f1-c90f-48c0-887e-d267f9085491' after 45 seconds

(...)


bosh -d cf vms

(...)

nats/e936cbd2-8d46-4099-bf4e-359be48d240c             unresponsive agent  z1  10.0.65.3   i-01543bd90121681f7  vm_2cpu_4gb   true    bosh-aws-xen-hvm-ubuntu-jammy-go_agent/1.199

```

I think it is because monit is technically part of the BOSH agent config and the job monit script gets copied over from the job to `/var/vcap/monit/jobs`. So if you put garbage in the monit script, it actually crashes the agent during startup and you can no longer ssh into the VM even:

```
Task 231257 | 11:21:40 | Error: Timed out sending 'ssh' to instance: 'nats/e936cbd2-8d46-4099-bf4e-359be48d240c', agent-id: 'bc8584f1-c90f-48c0-887e-d267f9085491' after 45 seconds
```

Looks like we need some strict checks on those limits...😅

domdom82 commented:
@geofffranks I've added code that fails the deployment if incompatible values are used for the limits. I've also added a spec test for all possible inputs. Please re-review, thanks!

geofffranks commented:
Another question for you on this - would it be more beneficial to have this limit autocalculated based on available memory on the VM? Otherwise scaling would mean increasing VM size, and then also increasing a bosh property.

mkocher commented:
I've seen one other report recently of NATS running out of memory, after not hearing much about NATS for a long time. Are you confident this is happening because of the scale of the environment? I'm wondering if there's a new NATS memory leak.

domdom82 commented:
@geofffranks this could be achieved by setting the limit to a percentage instead of a fixed number. e.g. `nats.mem_limit.alert = 50 %`
@mkocher that's interesting. we haven't seen an oom issue with our nats, yet an increase in memory usage overall that triggered this PR. We have over 60k apps now with the largest deployments so I'll attribute the increase in memory to that for now.
The issue we see and why we need to configure a higher limit:
1. NATS uses over 500M memory
2. Monit triggers alert
3. BOSH regards the job as failing
4. BOSH DNS takes this VM out of the list for `nats.services.cf.internal`
5. Other machines connect to the remaining VMs via DNS, leaving the seemingly "unhealthy" one without connections. 

This creates an imbalance in the NATS cluster, which is bad.

A second side-effect is that when CF gets deployed (even if there are no manifest changes), BOSH will automatically restart the failing (but not really failing) NATS job, further disrupting the environment as all existing connections to that NATS VM will be disrupted and have to reconnect again.

We will definitively look into the NATS cluster and do some profiling so see if there are mem leaks. However, we first need to stop the constant disruptions occurring by BOSH thinking the VM is unhealthy because the monit limit is too low.

mkocher commented:
👍 Thanks for the additional details. It's quite possible it's natural growth and not a memory leak.

winkingturtle-vmw commented:
@domdom82 what value are you setting for `memlimit` in your foundation and have you seen nats-server reach close to the new limit ?

domdom82 commented:
@winkingturtle-vmw we have set the limit to 50% on our landscapes and the average value is about 7.5% so we are far off the new limit atm. On smaller landscapes we have a smaller vm size where the memory usage has reached some 15% during peak times.
